### PR TITLE
[NV-1175] - Enforce environment param in dal query

### PR DESCRIPTION
**Pull request by djabarovgeorge**
### What change does this PR introduce? 

This PR will help us to make sure we won't forget to add environment variables to the DAL query.

### Why was this change needed?

Without adding environment variables to the DAL query we can fetch other users' data by mistake.

### CODE REVIEW

File: libs/dal/src/repositories/integration/integration.repository.ts [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+import { BaseRepository, Omit } from '../base-repository';
 import { IntegrationEntity } from './integration.entity';
 import { Integration } from './integration.schema';
 import { DalException } from '../../shared';
 
-export class IntegrationRepository extends BaseRepository<IntegrationEntity> {
+class PartialIntegrationEntity extends Omit(IntegrationEntity, ['_environmentId', '_organizationId']) {}
+
+type EnforceEnvironmentQuery = FilterQuery<PartialIntegrationEntity & Document> &
+  ({ _environmentId: string } | { _organizationId: string });
```
Comment (djabarovgeorge) [inline libs/dal/src/repositories/integration/integration.repository.ts:11]: Unfortunately, I could not create a more reusable piece of code. 
I will look into it later and hopefully will come up with more reusable code so we won't have these duplicated lines.

File: apps/api/src/app/events/usecases/process-subscriber/process-subscriber.usecase.ts [hunk 2]
```diff
@@ -72,8 +72,8 @@ export class ProcessSubscriber {
 
     for (const step of steps) {
       const integration = await this.integrationRepository.findOne({
-        organizationId: command.organizationId,
-        environmentId: command.environmentId,
+        _organizationId: command.organizationId,
+        _environmentId: command.environmentId,
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/process-subscriber/process-subscriber.usecase.ts:76]: :o So i guess it wasn't working here at all?

File: libs/dal/src/repositories/base-repository.ts [hunk 3]
```diff
@@ -105,11 +105,15 @@ export class BaseRepository<T> {
...
 
-  protected mapEntities(data: any): T[] {
-    return plainToInstance<T, T[]>(this.entity, JSON.parse(JSON.stringify(data)));
+  protected mapEntities(data: any): T_Response[] {
+    return plainToInstance<T_Response, T_Response[]>(this.entity, JSON.parse(JSON.stringify(data)));
   }
 }
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export const Omit = <T, K extends keyof T>(Class: new () => T, keys: K[]): new () => Omit<T, typeof keys[number]> =>
```
Comment (scopsy) [inline libs/dal/src/repositories/base-repository.ts:118]: Are you sure this is not provided globally by TS? From what I recall Omit is a utility available globally: 

https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys

File: libs/dal/src/repositories/integration/integration.repository.ts (hunk 1 above)
Comment (scopsy) [reply libs/dal/src/repositories/integration/integration.repository.ts:11]:
```
type EnforceEnvironmentQuery<T> = FilterQuery<T & Document> &
  ({ _environmentId: string } | { _organizationId: string });
```

Wonder if this syntax might be valid.

**Review by scopsy — APPROVED** [review]
Loooks great! Added a few comments. feel free to merge when you are ready :)

File: libs/dal/src/repositories/integration/integration.repository.ts (hunk 1 above)
Comment (p-fernandez) [reply libs/dal/src/repositories/integration/integration.repository.ts:11]: My take here would be to flip things around and create a `PartialIntegrationEntity` with all the content of `IntegrationEntity` but the `_environmentId` and `_organizationId` and then make:
```ts
class IntegrationEntity extends PartialIntegrationEntity {
   _environmentId: string;
   _organizationId: string;
```

So for the `EnforceEnvironmentQuery we can directly use:
```
type EnforceEnvironmentQuery = FilterQuery<PartialIntegrationEntity & Document>
```
What do you think?

**Comment by djabarovgeorge** [comment]
So there were a lot of things I did not take into account as the changes use-case and the environment cases etc.

I fixed the test but there may be cases that the test did not cover, so I would really like it if you could check the change especially 'job', 'changes', 'environment' 
@scopsy @ainouzgali @davidsoderberg

**Comment by scopsy** [comment]
@djabarovgeorge can we merge this one from your side?

File: apps/api/src/app/change/usecases/apply-change/apply-change.usecase.ts [hunk 4]
```diff
@@ -19,9 +19,7 @@ export class ApplyChange {
     });
 
     const changes = await this.changeRepository.find(
-      {
-        _parentId: parentChange._id,
-      },
+      { _environmentId: command.environmentId, _parentId: parentChange._id },
```
Comment (davidsoderberg) [inline apps/api/src/app/change/usecases/apply-change/apply-change.usecase.ts:22]: I think it is safer to take parentChanges env id 😄

Comment (djabarovgeorge) [reply apps/api/src/app/change/usecases/apply-change/apply-change.usecase.ts:22]: so the update will be 
      { _environmentId: parentChange._environmentId, _parentId: parentChange._id },
right?

**Pull request merged**